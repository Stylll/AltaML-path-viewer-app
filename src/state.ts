import { IDirection, IState } from './types';

export const direction: IDirection = {
    bound: {
        posXBound: [-2,-1,0,1,2,3],
        posYBound: [1,0,-1],
    },
    endPosition: {
        raw: [3,1],
        string: '3/1',
    },
    highestPosition: { x: 3,
        y: 1 },
    lastDirection: 'up',
    lowestPosition: { x: -2,
        y: -1 },
    multiOccurence: ['-2/-1'],
    path: {
        raw: [
            [0,0],
            [0,1],
            [-1,1],
            [-2,1],
            [-2,0],
            [-2,-1],
            [-1,-1],
            [0,-1],
            [1,-1],
            [1,0],
            [2,0],
            [3,0],
            [3,1],
        ],
        string: ['0/0','0/1','-1/1','-2/1','-2/0','-2/-1','-1/-1','0/-1','1/-1','1/0','2/0','3/0','3/1'],
    },
    startPosition: {
        raw: [0,0],
        string: '0/0',
    },
};

const directionNames: string[] = [
    'directions-1',
    'directions-2',
    'directions-3',
    'directions-4',
    'directions-5',
    'directions-6',
];

export const defaultState: IState = {
    direction: undefined,
    directionNames: directionNames,
};
