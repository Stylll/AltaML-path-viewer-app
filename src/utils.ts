export function fetchData (url: string): Promise<any> {
    return fetch(url)
        .then((response): Promise<any> => {
            return response.json();
        });
}
