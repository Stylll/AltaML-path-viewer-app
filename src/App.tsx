import React, { ReactElement, useState, SetStateAction, Dispatch, useEffect } from 'react';

// eslint-disable-next-line import/no-cycle
import { Footer } from './components/Footer/Footer.component';
import { GridMap } from './components/GridMap/GridMap.component';
import { Header } from './components/Header/Header.component';
import { AppContext } from './context';
import { defaultState } from './state';
import './App.css';
import { IDirection } from './types';
import { fetchData } from './utils';

const SERVER_API: string | undefined = process.env.REACT_APP_SERVER_API;

export function App (): ReactElement {

    const [direction, setDirection]:
    [IDirection | undefined, Dispatch<SetStateAction<IDirection | undefined>>]
    = useState<IDirection | undefined>(defaultState.direction);

    const [isLoading, setIsLoading]:
    [boolean, Dispatch<SetStateAction<boolean>>] = useState<boolean>(false);

    const [error, setError]:
    [string, Dispatch<SetStateAction<string>>] = useState<string>('');

    const errorMessage: string = 'An error occurred. Please try again.';

    function loadDirection (directionName: string): void {
        setIsLoading(true);
        setError('');

        const url: string = `${SERVER_API}?filename=${directionName}`;

        fetchData(url)
            .then((data): void => {
                setIsLoading(false);
                setDirection(data.data);
            })
            .catch((): void => {
                setIsLoading(false);
                setError(errorMessage);
            });
    }

    useEffect((): void => {
        loadDirection(defaultState.directionNames[0]);
    }, []);

    return (
        <AppContext.Provider value={
            {
                direction: direction,
                loadDirection: loadDirection,
                directionNames: defaultState.directionNames,
                isLoading: isLoading,
                error: error,
            }
        }
        >
            <div className="App">
                <Header />
                <GridMap />
                <Footer />
            </div>
        </AppContext.Provider>
    );
}
