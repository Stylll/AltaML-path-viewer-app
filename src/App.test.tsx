import { render, fireEvent, wait } from '@testing-library/react';
import React from 'react';

import { App } from './App';
import { direction } from './state';
import { fetchData } from './utils';

jest.mock('./utils');

const fetchDataMock = fetchData as jest.MockedFunction<typeof fetchData>;

describe('App Component', (): void => {

    afterEach((): void => {
        fetchDataMock.mockReset();
    });

    it('should render properly', async (): Promise<void> => {
        fetchDataMock.mockResolvedValue({
            data: direction,
        });

        const { baseElement } = render(<App />);

        await wait((): void => {
            expect(baseElement).toMatchSnapshot();
        });

        return;
    });

    it('should call fetchDataMock and display proper layout when it renders', async (): Promise<void> => {
        fetchDataMock.mockResolvedValue({
            data: direction,
        });

        const { container, getByText } = render(<App />);

        expect(container.querySelector('.loading')).toBeInTheDocument();
        await wait((): void => {
            expect(getByText('Path')).toBeInTheDocument();
        });
        expect(fetchDataMock).toHaveBeenCalled();
        return;
    });

    it('should display error message if error occurred when fetching', async (): Promise<void> => {
        fetchDataMock.mockRejectedValue({
            data: {
                message: 'An error occurred',
            },
        });

        const { container, getByText } = render(<App />);

        expect(container.querySelector('.loading')).toBeInTheDocument();
        fireEvent.click(getByText('Load Direction'));
        await wait((): void => {
            expect(getByText('An error occurred. Please try again.')).toBeInTheDocument();
        });
        expect(fetchDataMock).toHaveBeenCalledTimes(2);

        return;
    });
});
