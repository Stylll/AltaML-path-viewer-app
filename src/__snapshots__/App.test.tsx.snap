// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render properly 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="headerContainer"
      >
        <div
          class="headerContainer__titleSection"
        >
          <h1
            class="headerContainer__titleSection--title"
          >
            PathRenderer
          </h1>
          <h2
            class="headerContainer__titleSection--subtitle"
          >
            showing a generated path from a text direction
          </h2>
        </div>
        <div
          class="headerContainer__changeSection"
        >
          <span
            class="headerContainer__changeSection--selected"
          >
            directions-1
          </span>
          <div
            class="headerContainer__changeSection--options"
          >
            <p
              class="label"
            >
              select direction:
            </p>
            <select
              data-testid="directions"
              id="directions"
              name="directionName"
            >
              <option
                value="directions-1"
              >
                directions-1
              </option>
              <option
                value="directions-2"
              >
                directions-2
              </option>
              <option
                value="directions-3"
              >
                directions-3
              </option>
              <option
                value="directions-4"
              >
                directions-4
              </option>
              <option
                value="directions-5"
              >
                directions-5
              </option>
              <option
                value="directions-6"
              >
                directions-6
              </option>
            </select>
          </div>
          <div
            class="headerContainer__changeSection--button"
          >
            <button
              id="loadButton"
            >
              Load Direction
            </button>
          </div>
        </div>
      </div>
      <div>
        
        <div
          class="legendContainer"
        >
          <p
            class="legendContainer__legendTitle"
          >
            Legend
          </p>
          <div
            class="legendContainer__legendRow"
          >
            <div
              class="legendContainer__legendRow--col"
            >
              <div
                class="td active"
              />
              <span>
                Travel Path
              </span>
            </div>
            <div
              class="legendContainer__legendRow--col"
            >
              <div
                class="td multi"
              />
              <span>
                Cross Path
              </span>
            </div>
            <div
              class="legendContainer__legendRow--col"
            >
              <div
                class="td start"
              />
              <span>
                Start Position
              </span>
            </div>
            <div
              class="legendContainer__legendRow--col"
            >
              <div
                class="td stop"
              />
              <span>
                End Position
              </span>
            </div>
          </div>
        </div>
        <br />
        <div
          class="gridContainer"
        >
          <div>
            <p
              class="gridTitle"
            >
              Path
            </p>
            <table>
              <tbody>
                <tr>
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=""
                  />
                  <td
                    class=""
                  />
                  <td
                    class=" stop up"
                  />
                </tr>
                <tr>
                  <td
                    class=" active"
                  />
                  <td
                    class=""
                  />
                  <td
                    class=" start"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                </tr>
                <tr>
                  <td
                    class="multi"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=" active"
                  />
                  <td
                    class=""
                  />
                  <td
                    class=""
                  />
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <p>
          Stephen Aribaba Â© 2020
        </p>
      </div>
    </div>
  </div>
</body>
`;
